core/commonMain/src/implementations/immutableMap/PersistentHashMapBuilderContentIterators.kt

```diff
@@ -8,32 +8,37 @@ package kotlinx.collections.immutable.implementations.immutableMap
 import kotlinx.collections.immutable.internal.assert
 
 
-internal class TrieNodeMutableEntriesIterator<K, V>(private val builder: PersistentHashMapBuilder<K, V>)
-    : TrieNodeBaseIterator<K, V, MutableMap.MutableEntry<K, V>>() {
+internal class TrieNodeMutableEntriesIterator<K, V>(
+        private val parentIterator: PersistentHashMapBuilderEntriesIterator<K, V>
+) : TrieNodeBaseIterator<K, V, MutableMap.MutableEntry<K, V>>() {
 
     override fun next(): MutableMap.MutableEntry<K, V> {
         assert(hasNextKey())
         index += 2
         @Suppress("UNCHECKED_CAST")
-        return MutableMapEntry(builder, buffer[index - 2] as K, buffer[index - 1] as V)
+        return MutableMapEntry(parentIterator, buffer[index - 2] as K, buffer[index - 1] as V)
     }
 }
 
-private class MutableMapEntry<K, V>(private val builder: PersistentHashMapBuilder<K, V>,
-                                    key: K,
-                                    override var value: V) : MapEntry<K, V>(key, value), MutableMap.MutableEntry<K, V> {
+private class MutableMapEntry<K, V>(
+        private val parentIterator: PersistentHashMapBuilderEntriesIterator<K, V>,
+        key: K,
+        override var value: V
+) : MapEntry<K, V>(key, value), MutableMap.MutableEntry<K, V> {
+
     override fun setValue(newValue: V): V {
         val result = value
         value = newValue
-        builder[key] = newValue
```

Comment:
Here is the bug https://github.com/Kotlin/kotlinx.collections.immutable/issues/68.
The value is updated bypassing the iterator this mutable entry was obtained from.